com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java

```diff
@@ -1800,8 +1805,14 @@ private List<TaskResult> runTasks(IProgress monitor) throws IOException {
                     }
                 }
                 TimeProfiler.stop();
-
-                boolean shouldRun = (!allOutputExists || !allSigsEquals) && !completedTasks.contains(task);
+                // game.project is always the last task
+                boolean isLastTask = completedTasks.size() + 1 == buildTasks.size();
+                boolean shouldRun = !completedTasks.contains(task);
+                // GameProjectBuilder creates archives, and it is always the last task.
+                // If for some reason it's not, something went wrong, and the build pipeline is broken.
+                // But some tests may run build for some particular files without building game.project at all.
+                shouldRun = shouldRun && (!allOutputExists || !allSigsEquals ||
+                            (isLastTask && buildContainsChanges && (task.getBuilder() instanceof GameProjectBuilder)));
```

Comment:
Check task type, but to avoid checking every single task, use it together with `isLastTask`

Reply:
@britzl the approach with Exception didn't work, because some tests build separate files.
Instead, I did the change you initially asked but combine it with `isLastTask` to be sure it work fast on big projects

Reply:
One possible tiny improvement could be to have a isGameProjectBuilder() on the Builder class, default to false and override to true for GameProjectBuilder.